### PR TITLE
Fstr sage

### PR DESCRIPTION
**Pull request by igormarkov00**
Implemented draft version of SAGE values calculation. 
This version works only for datasets with no groupping and only float features. 
Parameters of algorithm are **not** tuned, they were set empirically for querywise dataset and may not perform well for any others.

### CODE REVIEW

File: catboost/libs/fstr/sage_values.cpp [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+    FullSubsetIndexingPtrWrapper() = default;
+
+    TAtomicSharedPtr<TArraySubsetIndexing<ui32>> Get(ui32 objectsCount) {
+        CB_ENSURE(objectsCount, "Size of TArraySubsetIndexing should be greater than 0");
+        if (auto fullSubsetIndexingPtr = vSubsetIndexing[objectsCount]) {
+            return fullSubsetIndexingPtr;
+        }
+        return vSubsetIndexing[objectsCount] = MakeAtomicShared<TArraySubsetIndexing<ui32>>(
+                                                   TFullSubset<ui32>{objectsCount}
+                                               );
```
Comment (ek-ak) [inline catboost/libs/fstr/sage_values.cpp:43]: скобка уехала

File: catboost/libs/fstr/sage_values.cpp [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+
+using namespace NCB;
+
+// class for TArraySubsetIndexing pointers long-living
+class FullSubsetIndexingPtrWrapper {
+public:
+    FullSubsetIndexingPtrWrapper() = default;
+
+    TAtomicSharedPtr<TArraySubsetIndexing<ui32>> Get(ui32 objectsCount) {
+        CB_ENSURE(objectsCount, "Size of TArraySubsetIndexing should be greater than 0");
```
Comment (ek-ak) [inline catboost/libs/fstr/sage_values.cpp:37]: CB_ENSURE_INTERNAL

File: catboost/libs/fstr/sage_values.cpp [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+double CalculateModelLoss(
+    const TFullModel& model,
+    const TDataProvider& dataset,
+    const TVector<THolder<IMetric>>& metrics,
+    NPar::ILocalExecutor* localExecutor)
+{
+    TMetricsPlotCalcer estimator(
+        model,
+        metrics,
+        "",  // fictive parameter
```
Comment (ek-ak) [inline catboost/libs/fstr/sage_values.cpp:302]: `/*tmpDir*/ "",`

File: catboost/libs/fstr/sage_values.cpp [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+        model.GetTreeCount() - 1,
```
Comment (ek-ak) [inline catboost/libs/fstr/sage_values.cpp:303]: только по последнему дереву?

File: catboost/libs/fstr/sage_values.cpp [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+        model.GetTreeCount(),
+        1,
```
Comment (ek-ak) [inline catboost/libs/fstr/sage_values.cpp:305]: `/*step*/ 1,`

File: catboost/libs/fstr/sage_values.cpp [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+        1,
```
Comment (ek-ak) [inline catboost/libs/fstr/sage_values.cpp:306]: `/*processIterationStep*/ 1,`

File: catboost/libs/fstr/sage_values.cpp [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+    }
+}
+
+double CalculateModelLoss(
+    const TFullModel& model,
+    const TDataProvider& dataset,
+    const TVector<THolder<IMetric>>& metrics,
+    NPar::ILocalExecutor* localExecutor)
+{
+    TMetricsPlotCalcer estimator(
```
Comment (ek-ak) [inline catboost/libs/fstr/sage_values.cpp:299]: use `CreateMetricCalcer` function

File: catboost/libs/fstr/sage_values.cpp [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+    size_t nSamples,
+    size_t batchSize)
+{
+    CB_ENSURE(model.ModelTrees->GetDimensionsCount() == 1, "Model must not be trained for multiclassification");
+
+    // setting algorithm params
+    size_t featuresCount = dataset.MetaInfo.GetFeatureCount();
+    batchSize = Min(batchSize, size_t(dataset.GetObjectCount()));
+    auto featuresLayout = dataset.MetaInfo.FeaturesLayout;
+    // Cout << "Objects Count: " << dataset.GetObjectCount() << Endl;
```
Comment (ek-ak) [inline catboost/libs/fstr/sage_values.cpp:372]: лишнее

File: catboost/python-package/ut/medium/catboost-python-package-ut-medium [hunk 9]
```diff
@@ -0,0 +1 @@
+/home/ds4xts/.ya/build/symres/83d0cbbbbe956b11113838380f69a9d6/catboost-python-package-ut-medium
```
Comment (ek-ak) [inline catboost/python-package/ut/medium/catboost-python-package-ut-medium:1]: лишнее

File: catboost/python-package/ut/medium/python_binary/catboost-python [hunk 10]
```diff
@@ -0,0 +1 @@
+/home/ds4xts/.ya/build/symres/e7947c4db1e4161311dd07d070a3f488/catboost-python
```
Comment (ek-ak) [inline catboost/python-package/ut/medium/python_binary/catboost-python:1]: лишнее

File: catboost/python-package/ut/medium/test.py [hunk 11]
```diff
@@ -4023,6 +4023,17 @@ def test_feature_importance_prettified(task_type):
     return local_canonical_file(fimp_txt_path)
 
 
+def test_feature_importance_sage(task_type):
```
Comment (ek-ak) [inline catboost/python-package/ut/medium/test.py:4026]: task_type здесь не нужен

File: catboost/python-package/ut/medium/test.py [hunk 12]
```diff
@@ -4023,6 +4023,17 @@ def test_feature_importance_prettified(task_type):
...
+    pool = Pool(TRAIN_FILE, column_description=CD_FILE)
+    model = CatBoostClassifier(iterations=5, learning_rate=0.03, task_type=task_type, devices='0')
+    model.fit(pool)
+
+    feature_importances = model.get_feature_importance(pool, type=EFstrType.SageValues, sage_n_samples=8, sage_batch_size=128)
+    fimp_npy_path = test_output_path(FIMP_NPY_PATH)
+    np.save(fimp_npy_path, np.array(feature_importances))
+    return local_canonical_file(fimp_npy_path)
+
+
```
Comment (ek-ak) [inline catboost/python-package/ut/medium/test.py:4037]: еще нужен тест в pytest
https://github.com/catboost/catboost/blob/master/catboost/pytest/test.py

**Review by ek-ak — CHANGES_REQUESTED** [review]

File: catboost/libs/fstr/sage_values.cpp (hunk 8 above)
Comment (igormarkov00) [reply catboost/libs/fstr/sage_values.cpp:372]: Уберу позже. Пока удобно для отладки

File: catboost/libs/fstr/sage_values.cpp (hunk 4 above)
Comment (igormarkov00) [reply catboost/libs/fstr/sage_values.cpp:303]: Да

**Comment by igormarkov00** [comment]
Зафиксировать random seed

File: catboost/libs/fstr/sage_values.cpp [hunk 13]
```diff
@@ -0,0 +1,470 @@
...
+
+class MarginalImputer {
+public:
+    MarginalImputer(
+        const TDataProvider& dataset,
+        NPar::ILocalExecutor* executor,
+        ui64 randomSeed = 228)
+        : rand(randomSeed)
+    {
+        localExecutor = executor;
```
Comment (ek-ak) [inline catboost/libs/fstr/sage_values.cpp:57]: CammelCase for class variables

File: catboost/libs/fstr/sage_values.cpp [hunk 14]
```diff
@@ -0,0 +1,473 @@
...
+                    break;
+                }
+
+                default:
+                    CB_ENSURE_INTERNAL(false, "Unknown feature type");
+            }
+        }
+    }
+
+    TDataProvider Impute(
```
Comment (ek-ak) [inline catboost/libs/fstr/sage_values.cpp:195]: This function not used anywhere, delete it

File: catboost/libs/fstr/sage_values.cpp [hunk 15]
```diff
@@ -0,0 +1,473 @@
...
+        ui32 embeddingFeaturesCount = dataset.MetaInfo.FeaturesLayout->GetEmbeddingFeatureCount();
+        for (ui32 featureIndex = 0; featureIndex < embeddingFeaturesCount; ++featureIndex) {
+            auto featureValues = (*referenceDataProviderPtr->GetEmbeddingFeature(featureIndex))->
+                                 ExtractValues(LocalExecutor);
+            ReferenceSamplesEmbedding.push_back(std::move(featureValues));
+        }
+    }
+
+    void ImputeInplace(
+        TDataProvider& dataset,
```
Comment (ek-ak) [inline catboost/libs/fstr/sage_values.cpp:98]: Not constant args are provided by pointer and after constant args

File: catboost/libs/fstr/sage_values.cpp [hunk 16]
```diff
@@ -0,0 +1,473 @@
...
+        const TVector<std::pair<ui32, EFeatureType>>& features,
```
Comment (ek-ak) [inline catboost/libs/fstr/sage_values.cpp:99]: used only 1-element vectors, so no need to wrap in vector

**Review by ek-ak — CHANGES_REQUESTED** [review]

**Comment by ek-ak** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@igormarkov00, please agree to the CLA (https://github.com/catboost/catboost/blob/master/CONTRIBUTING.md#provide-contributions)

**Comment by arcadia-devtools** [comment]
CLA already signed

**Comment by ek-ak** [comment]
@arcadia-devtools Ship it!

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
@arcadia-devtools Ship it!

**Comment by ek-ak** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@ek-ak, internal review request created: [2535035](https://a.yandex-team.ru/review/2535035)
